### ISSUE TITLE
Automatic Windows builds?

### ISSUE THREAD

**Issue by roy7** [leela-zero/leela-zero#2226]
@gcp Any change to the AppVeyor situation with automated Windows builds? Hopefully OpenBLAS/Nuget has been updated since a year ago? Last update I saw on this was Feb 2018 when we couldn't do it since Nuget was outdated.

**Comment by l1t1** [comment]
where can download the win build of next branch

**Comment by roy7** [comment]
Looking around the Nuget web site it appears noone has updated the packages still. I'll try contacting the creators of both existing packages to see if they'll provide a 0.3.5 one.

**Comment by roy7** [comment]
I have no idea what sort of math we're using BLAS for, but looking at https://medium.com/datathings/benchmarking-blas-libraries-b57fb1c6dc7 it seems for smaller size calculations Eigen is fastest. At larger sizes cuBLAS is fastest.

I was looking into this out of curiosity, since Nuget does have a seemingly up to date https://www.nuget.org/packages/Tensor/ package, from http://www.deepml.net/Tensor/, which uses cuBLAS.

**Comment by roy7** [comment]
Actually I found this.

https://github.com/leela-zero/leela-zero/pull/1900

Are binaries automatically being made already and I just have no idea where to look to find them? I thought latest might appear on the Releases page but they don't, so that's why I opened this issue.

**Comment by roy7** [comment]
Not something I feel comfortable experimenting with, but if we're using Eigen and it's safe to publish binaries from AppVeyor, we'd set up a Github deployment following https://www.appveyor.com/docs/deployment/github/. I wonder if we can have a single "nightly" release (and not store every single build on the releases page).

Also it seems current Leela Zero updates push to https://ci.appveyor.com/project/gcp/leela-zero-8arv1 instead of https://ci.appveyor.com/project/gcp/leela-zero so the README and possibly our Github config might need updated?

**Comment by l1t1** [comment]
rn has a build download link
https://github.com/zakki/Ray/issues/171
CI artifact is available. https://ci.appveyor.com/project/zakki/ray/builds/21043153/artifacts

**Comment by gcp** [comment]
The Leela Chess Zero guys told me at some point they had this set up (automatic OpenBLAS build), but I can't find the thread. Maybe @borg323 remembers?

**Comment by gcp** [comment]
>if we're using Eigen and it's safe to publish binaries from AppVeyor

Eigen binaries only work on the platform they're compiled on. They're not suitable for releases.

**Comment by borg323** [comment]
No, we haven't got automatic OpenBLAS builds yet, but I'll be happy to set it up if you can give me some hints (compiler flags etc).

**Comment by roy7** [comment]
I haven't gotten a reply from either maintainer I asked to update the OpenBLAS packages. I'd need to maybe submit my own, or would we want to subproject it like Eigen?

Or can Appveyor build script auto pull down OpenBLAS to build it fresh instead of relying on the packages?

**Comment by gcp** [comment]
>No, we haven't got automatic OpenBLAS builds yet, but I'll be happy to set it up if you can give me some hints (compiler flags etc).

make TARGET=HASWELL DYNAMIC_ARCH=1 NUM_THREADS=64 NO_LAPACK=1 USE_THREAD=1 ONLY_CBLAS=1

I also manually patched Makefile.rule and Makefile.system to remove (-D)USE_TLS.

**Comment by roy7** [comment]
Can AppVeyor create two binaries, like one for TARGET=HASWELL and one for TARGET=ZEN, or is that a meaningless optimization since our bottleneck is always GPU evals and never cpu?

**Comment by roy7** [comment]
I looked around the OpenBLAS web site a bit and they said binary releases are available automatically from SourceForge. Looks like https://sourceforge.net/projects/openblas/files/latest/download is hard linked to the latest version at all times.

If I understand it, the files in the kernel/ folder (all .S files) are the assembly source code which is why they call it a "binary" release? I was expecting to find a .DLL or something.

I don't know if those can be used on-demand to link into an AppVeyor build so we don't need to rely on NuGet for a package, since noone has updated those in 2 years.

**Comment by borg323** [comment]
I'm working on the appveyor build - at the moment the DYNAMIC_ARCH=1 build takes more than an hour, which is a hard limit for the free appveyor offering.

**Comment by gcp** [comment]
>Can AppVeyor create two binaries, like one for TARGET=HASWELL and one for TARGET=ZEN, or is that a meaningless optimization since our bottleneck is always GPU evals and never cpu?

The TARGET doesn't matter if DYNAMIC_ARCH=1 is set. It will include all targets and select at startup.

**Comment by roy7** [comment]
Oh, perfect.

**Comment by borg323** [comment]
Took a bit longer than I expected, but here it is: https://github.com/borg323/OpenBLAS/blob/bin_for_leela/appveyor.yml
I changed NUM_THREADS to 128 (like the 0.3.3 version from sjeng.org) and added LDFLAGS=-static in order to avoid an additional dll for exception handling.
The 64 bit version takes very long to build, depending on the vm used it may not finish in the 1 hour allowed by appveyor - seems to be OK outside of US business hours.

Test binaries: https://ci.appveyor.com/api/buildjobs/rkkdeslssm72gjdn/artifacts/openblas_64.zip and https://ci.appveyor.com/api/buildjobs/v1ht1kfso2b254qp/artifacts/openblas_32.zip (64 and 32 bit respectively).

**Comment by l1t1** [comment]
thanks

**Comment by roy7** [comment]
The binaries are just the openblas.dll, there's no leelaz.exe. Do I misunderstand how this works? :)

**Comment by roy7** [comment]
Do I need to merge that appveyor.yml into AppVeyor or into LZ's github? I'm unsure what to do. Or maybe gcp will handle it.

**Comment by borg323** [comment]
The idea is to get an openblas zip from a known location - probably a release from my fork. This is how I do it for lc0: https://github.com/LeelaChessZero/lc0/blob/master/appveyor.yml#L20

For the time being a few tests replacing the dll in existing installations would be welcome. Seemed to be fine with lc0, both 32 and 64 bit versions, but I only tried on an old core2 laptop.

**Comment by roy7** [comment]
Oh ok right I understand. It's like replacing the need for an updated Nuget package, we just use your pre-compiled OpenBLAS stuff instead to make the Appveyor auto-compilation faster?

Would it be simpler perhaps for you or someone to just submit a newer OpenBLAS package to Nuget?

**Comment by borg323** [comment]
This is a very stripped down build with only BLAS, so probably not appropriate for Nuget. And it is easy for anyone to take it over if I disappear.

**Comment by roy7** [comment]
Ok cool thank you. :)

**Comment by borg323** [comment]
Are there any comments regarding the packaging of the zip files? If not I'll proceed with a github "release" to have permanent links for the files - the appveyor ones will expire in 5 months.
I intend to create fresh binaries every time I notice a new OpenBLAS release, but I'm not watching them very closely.
